#!/usr/bin/env python2

from src.panda_moveit import PandaMoveitInterface
import sys, rospy, moveit_commander
from geometry_msgs.msg import Pose
import copy
import os


def generateWaypointFromLine(line, panda):
    # Get target_pose and transform tcp->gym
    target_pose = list()
    for e in line:
        target_pose.append(float(e))
    target_wrist = panda.getWristFromTCP(target_pose[:7])

    # Generate waypoint and add it
    waypoint = Pose()
    waypoint.position.x = target_wrist[0]
    waypoint.position.y = target_wrist[1]
    waypoint.position.z = target_wrist[2]
    waypoint.orientation.x = target_wrist[3]
    waypoint.orientation.y = target_wrist[4]
    waypoint.orientation.z = target_wrist[5]
    waypoint.orientation.w = target_wrist[6]
    return waypoint


def main(FILE_PATH):
    try:
        # Initialize moveit_commander and rospy
        moveit_commander.roscpp_initialize(sys.argv)
        rospy.init_node('panda_trajectory', anonymous=True)

        # Get roslaunch parameters
        use_real_robot = rospy.get_param('~use_real_robot', False)
        gripper_speed = rospy.get_param('~gripper_speed', False)
        grasp_force = rospy.get_param('~grasp_force', False)
        grasp_epsilon_inner = rospy.get_param('~grasp_epsilon_inner', False)
        grasp_epsilon_outer = rospy.get_param('~grasp_epsilon_outer', False)

        # Create panda moveit interface
        panda = PandaMoveitInterface(delay=1, gripper_enabled=use_real_robot)

        # Read trajectory file
        with open(FILE_PATH, 'r') as file_reader:
            # GRIPPER OPEN
            line = file_reader.readline()
            gripper_width = float(line)
            panda.gripper_move(gripper_width, gripper_speed)


            # START
            line = file_reader.readline()
            tcp_target_pose = list()
            for e in line.split():
                tcp_target_pose.append(float(e))
            panda.moveToArmPoseTCP(tcp_target_pose)


            # PRE-GRASP and GRASP
            waypoints = list()
            while True:
                line = file_reader.readline()
                components = line.split()
                
                if len(components) == 1:
                    break

                waypoint = generateWaypointFromLine(components, panda)
                waypoints.append(copy.deepcopy(waypoint))
            panda.execute_cartesian_path(waypoints=waypoints)


            # CLOSE GRIPPER
            gripper_width = float(line)
            panda.gripper_grasp(gripper_width, grasp_epsilon_inner, grasp_epsilon_outer, gripper_speed, grasp_force)


            # PLACE
            waypoints = list()
            while True:
                line = file_reader.readline()
                components = line.split()
                
                if len(components) == 1:
                    break

                waypoint = generateWaypointFromLine(components, panda)
                waypoints.append(copy.deepcopy(waypoint))
            panda.execute_cartesian_path(waypoints=waypoints)


            # OPEN GRIPPER
            gripper = float(line)
            panda.gripper_move(gripper_width, gripper_speed)


            # Here you can read the last value to have the perfect alignment with panda_gym. This value is generated by a null action and therefore corresponds to the last current_pose read in panda_gym
            # line = file_reader.readline()
            # tcp_target_pose = list()
            # for e in line.split():
            #     tcp_target_pose.append(float(e))
            # panda.moveToArmPoseTCP(tcp_target_pose)


    except rospy.ROSInterruptException:
        print("ROS interrupted")



if __name__ == "__main__":
    FILE_NAME = "trajectory_test"


    file_path = os.path.join(os.path.dirname(__file__), "../data/trajectories/" + FILE_NAME + ".txt")
    main(file_path)